src/solver/type_getters.jl

```diff
@@ -477,6 +482,42 @@ thermo_state_type(::EquilMoistModel, ::Type{FT}) where {FT} = TD.PhaseEquil{FT}
 thermo_state_type(::NonEquilMoistModel, ::Type{FT}) where {FT} =
     TD.PhaseNonEquil{FT}
 
+function get_restart_file(config::AtmosConfig, base_output_dir)
```

Comment:
I tried refactoring this a bit, and I think there may be a missing edge case?

```julia
to_named_tuple(dict::Dict) = (; (Symbol(k) => v for (k, v) in dict)...)
get_restart_file(config::AtmosConfig, base_output_dir) =
    get_restart_file(base_output_dir; to_named_tuple(config.parsed_args)...)

function get_restart_file(base_output_dir; detect_restart_file=nothing, output_dir_style=nothing, restart_file=nothing)
    isnothing(restart_file) || return restart_file
    lowercase(output_dir_style) == "activelink" ||
        error("detect_restart_file works only with ActiveLink")
    # output_dir will be something like ABC/DEF/output_1234
    name_rx = r"output_(\d\d\d\d)"
    restart_file_rx = r"day\d+\.\w+\.hdf5"
    existing_outputs =
        filter(x -> !isnothing(match(name_rx, x)), readdir(base_output_dir))
    if !isempty(existing_outputs)
        latest_output = first(sort(existing_outputs, rev = true))
        previous_folder = joinpath(base_output_dir, latest_output)
        possible_restart_files = filter(
            f -> occursin(restart_file_rx, f),
            readdir(previous_folder),
        )
        if !isempty(possible_restart_files)
            restart_file_name =
                last(CA.sort_files_by_time(possible_restart_files))
            return joinpath(previous_folder, restart_file_name)
        end
    else
        error("??")
    end
end
```

This function no longer depends on `parsed_args`

Reply:
`restart_file` is set to `nothing` at the beginning of the function and changed only when `!isempty(existing_outputs)`. Having an empty existing outputs is allowed: it means that it is the first time we run this simulation.

Reply:
This rewrite is string-parsing with extra steps and it is the only function in type getters to be structured like this. If we want to do this properly and really decouple the function from argument parsing, we should use the correct type for `output_dir_style`, a subtype of `ClimaUtilities.OutputPathGenerator.OutputPathGeneratorStyle`, instead of matching a string.

Reply:
I changed to not use `parsed_args` and dispatch on the type of `output_dir_style`